tls: simplify implementation and fix one class of crashing bug

mattklein123 opened the pull request:
This change does 2 things:
1) It greatly simplifies the TLS implementation by removing the
bookkeeper code. The main insight is that the slot itself does not
need to be captured in any callbacks, just the index. By doing this,
all bookkeeper complexity can be removed and all callbacks work
directly with the slot index which can be immediately recycled when
the slot is deleted.
2) Adds a "still alive" shared_ptr which is captured by weak_ptr in
all callbacks. This does not completely prevent broken captures,
but it does fix the common case of a slot being immediately deleted
before any callbacks run on the workers (which can be common during
initial startup listener creation failure for example).

In a follow up change I will look into adding a clang-tidy check to prevent
capturing "this" in any TLS callback.

Fixes https://github.com/envoyproxy/envoy/issues/12364

Risk Level: Medium
Testing: TBD
Docs Changes: N/A
Release Notes: N/A

mattklein123 commented:
@jmarantz @htuch @ggreenway @stevenzzzz @lambdai this is seriously lacking tests but fixes https://github.com/envoyproxy/envoy/issues/12364 and I think would prevent a bunch of other issues we have had and have done one off fixes for (some of those fixes can likely be reverted).

Anyway, before I invest a huge amount of time I wanted to get some initial feedback here. Let me know what you think about this direction. I tried to add a bunch of comments in the code but lmk if I need to explain more. After the community meeting convo it should be pretty clear what is going on.

mattklein123 commented on source/extensions/access_loggers/grpc/http_grpc_access_log_impl.h:
```diff
@@ -51,12 +51,12 @@ class HttpGrpcAccessLog : public Common::ImplBase {
   std::vector<Http::LowerCaseString> response_headers_to_log_;
   std::vector<Http::LowerCaseString> response_trailers_to_log_;
   std::vector<std::string> filter_states_to_log_;
+  const ThreadLocal::SlotPtr tls_slot_; // Destruct first to flush callbacks.
```
I think this type of change can likely be removed via also deleting the still alive guard so it can't be locked. Then the callbacks would flush out but wouldn't actually be executed. Although, now that I'm typing this, I'm wondering if all of this can be simplified to just have the callbacks not run at all once the slot goes away. I will think about this more tomorrow.

mattklein123 replied:
Oh nevermind, I know why I didn't do this, because the shared_ptr deletion can race within the lambda, and there is still a problem.

stevenzzzz commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
+  // See the constructor for why we only do this in the running state.
+  if (parent_.state_ == State::Running) {
+    absl::MutexLock lock(&shutdown_mutex_);
+    shutdown_mutex_.Await(absl::Condition(&ready_to_destroy_));
```
is there a deadlock risk if a worker thread accidentally posts the wrapped callback back to main thread?

mattklein123 replied:
In practice no, since the worker can't actually get access to the wrapped callback, only the unwrapped one.

stevenzzzz replied:
I see.
Another risk though a long event in worker thread may block the main thread for a long while.
 
can we combine this active checking together with the "deferred deletion" trick?

I actually dont know why the bookkeeper trick would fail in the first place. that trick tries to guarantee the liveness of the slot when there are flying cb in worker threads as well.

mattklein123 replied:
> Another risk though a long event in worker thread may block the main thread for a long while.

Yes, this is possible. In practice, I'm not sure this matters since slot deletion is a rare operation and the workers should not be blocking, and we also have deadlock detection.

> can we combine this active checking together with the "deferred deletion" trick?

What do you mean?

> I actually dont know why the bookkeeper trick would fail in the first place. that trick tries to guarantee the liveness of the slot when there are flying cb in worker threads as well.

I explained this here: https://github.com/envoyproxy/envoy/issues/12364#issuecomment-675631096. Bookkeeper does not fix the issue in which people capture something in the lambda, and that thing gets destroyed before the callback runs. This is the bug in the linked issue and have had this bug at least 2-3 times before, it comes up quite often. With that said, it's possible we could get a long way with something like this: https://github.com/mesos/clang-tools-extra/blob/mesos_90/clang-tidy/mesos/ThisCaptureCheck.cpp which @jpeach sent me.

stevenzzzz replied:
> > Another risk though a long event in worker thread may block the main thread for a long while.
> 
> Yes, this is possible. In practice, I'm not sure this matters since slot deletion is a rare operation and the workers should not be blocking, and we also have deadlock detection.
> 
I have some concern here, the lock may take enough long to impact XDS or other OP on main thread, whe we have dozens of workers. for example, a RDS Config provider holds a tls slot, in SRDS we dynamically add or delete the 
RDS providers in each SRDS reponse. This could potentially put a performance hit on main thread.

> > can we combine this active checking together with the "deferred deletion" trick?
> 
> What do you mean?
> 
I mean, instead of hanging main thread, we could potentially put this slot into a deferred deletion queue, and schedule a deletionCB in main thread to cleanup all the slot instance when the ready_to_destroy_ bit flipped to true.

> > I actually dont know why the bookkeeper trick would fail in the first place. that trick tries to guarantee the liveness of the slot when there are flying cb in worker threads as well.
> 
> I explained this here: [#12364 (comment)](https://github.com/envoyproxy/envoy/issues/12364#issuecomment-675631096). Bookkeeper does not fix the issue in which people capture something in the lambda, and that thing gets destroyed before the callback runs. This is the bug in the linked issue and have had this bug at least 2-3 times before, it comes up quite often. With that said, it's possible we could get a long way with something like this: https://github.com/mesos/clang-tools-extra/blob/mesos_90/clang-tidy/mesos/ThisCaptureCheck.cpp which @jpeach sent me.

mattklein123 replied:
> I mean, instead of hanging main thread, we could potentially put this slot into a deferred deletion queue, and schedule a deletionCB in main thread to cleanup all the slot instance when the ready_to_destroy_ bit flipped to true.

As I just explained to you in DM, this doesn't actually fix the problem. We really have only 2 options:
1) We have to prevent the problem (blocking)
2) We have to fix all call sites (probably clang-tidy plugin, etc.)

lambdai replied:
My understanding is that the crash is only possible when worker has not started, 
But the cost at conditional wait is applied to working running state?! `parent_.state_ == State::Running`

mattklein123 replied:
It's not just before workers have started. The issue also happens post worker start if a listener fails to get fully constructed for example during LDS update.

jmarantz replied:
My main suggestion here is that, in case something goes wrong (deadlock or just slow exit), use `AwaitWithTimeout` in a loop, logging `info` every N seconds so that an operator knows why we're stuck even if we don't have debug signals. Ideally we could log progress toward being able to move on.

In the best case that log gets printed very few times, and in the worst case it's very noisy but we'll have an easier time diagnosing.

mattklein123 replied:
Yeah, I did consider also just always doing a timeout to make sure we don't deadlock, or do what you suggest with a limit.

lambdai replied:
> It's not just before workers have started. The issue also happens post worker start if a listener fails to get fully constructed for example during LDS update.

Right, anytime a slot is destroyed before the worker thread play with it.

Can we enforce removing the slot destroy *after* worker thread drop reference? E.g. always wrap an "attempt to delete function" posted to master thread regardless user has the `complete_cb`?

mattklein123 replied:
> Can we enforce removing the slot destroy after worker thread drop reference? E.g. always wrap an "attempt to delete function" posted to master thread regardless user has the complete_cb?

This is the way it works today. :) This is "bookkeeper". The problem is this method does not actually fix the dangling lambda capture issue. You simply can't fix it without blocking.

So like I said above, I think we have to decide:
1) Blocking is OK
2) Blocking is not OK and we should look into clang-tidy plugins or similar.

ahedberg replied:
Is mismanagement of object lifetimes the cause of the underlying bugs which this PR is trying to fix? If so, I'm not sure adding code complexity and blocking to paper over the issue is a good idea. I'd be much more in favor of seeing how far clang-tidy plugins get us.

mattklein123 replied:
> Is mismanagement of object lifetimes the cause of the underlying bugs which this PR is trying to fix?

Yes. Though, to be fair, the existing code already does this (see Bookkeeper).

> If so, I'm not sure adding code complexity and blocking to paper over the issue is a good idea. I'd be much more in favor of seeing how far clang-tidy plugins get us.

Yes that's a fair viewpoint and I'm fine if that is the group consensus.

ahedberg replied:
I probably would have voted against bookkeeper too, FWIW. In general, if a user is doing something incorrectly, I'd rather we surface that information to the user so they can implement a fix on their end that makes sense, rather than trying to detect that they made a mistake and prevent it from causing problems. The latter feels more bug-prone to me.

mattklein123 replied:
@ahedberg at a high level I agree with you. I'm torn on this one though, as this is a really thorny issue that has been the cause of many different crashes. The issue here is that lifetime alone is not necessarily a complete fix. Meaning, depending on how someone implements the captures, there can still be race conditions with deletion. Let me summarize the current state opinions below and we can think about a path forward.

stevenzzzz replied:
see https://github.com/envoyproxy/envoy/issues/7902 we will need either the bookkeeper or the change in this PR to fix that issue.

ggreenway commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
+  // Also capture still_alive_guard_ so we can verify the slot is still alive as it might have
+  // been destroyed pre-running state. See SlotImpl().
+  return [active_callback_guard = active_callback_guard_,
+          still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
```
Design alternative: what if this was a strong_ptr (or just use a refcnt at that point), and always allow the `cb` to run, to simplify reasoning about tls callbacks?

mattklein123 replied:
Can you explain more? I don't follow.

ggreenway replied:
Nevermind

mattklein123 commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
+            //    when a slot was destroyed before we entered the running state (during xDS init).
+            //    This is guaranteed by the blocking in slot destruction. Thus, no further
+            //    synchronization is needed.
+            if (parent.state_ == State::Running && still_alive_guard.lock()) {
```
I can add more comments, but it's not actually required. Basically, the weak_ptr/shared_ptr dance is not used to serialize across threads. Once threading starts, we use the lock (in the running state). This is used to serialize the case in which the slot was deleted *before* we started running.

ggreenway replied:
Oh, I see what you mean. Please add comments because it looks wrong at first glance.

lizan replied:
If you don't need the return value, you can just check `still_alive_guard.expired()`

mattklein123 commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
+  // been destroyed pre-running state. See SlotImpl().
+  return [active_callback_guard = active_callback_guard_,
+          still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
+    if (still_alive_guard.lock()) {
```
See above. Same reason.

lambdai replied:
@mattklein123  Could you use `!still_alive_guard.expired()`? expired() is straightforward when the goal is to fire a check.

mattklein123 replied:
Yes if we stick with this I will switch to expired()

mattklein123 commented:
> Are we creating different, but still hard-to-diagnose issues, by possibly not running the callback at all (if the slot was destroyed)?

The default implementation does run the callback. It's not run in the immediate deletion case on startup. IMO it's fine to not run the callback if the slot has been deleted so I think this one is OK.

> What's the perf impact? An additional atomic operation for each callback? If it's on all workers, it will probably be contended (in hardware). A wrapper function around the pass-in function.

Yes some additional atomic increments for each callback. There is no blocking on worker updates, but there could be contention around the atomics. With that said, I think contention on stats would be vastly worse, so I don't view this as a large issue.

> What's the worst-case wait time? (But it's only the main thread that waits, right?)

Worst case wait time on destruction is the event loop delay across all workers which could be hundreds of milliseconds. This is certainly the part that is not great and I'm happy to abandon this approach if we don't want this. I think blocking is the only way to fix the underlying problem if we don't take a different path and do something with code analysis like here: https://github.com/mesos/clang-tools-extra/blob/mesos_90/clang-tidy/mesos/ThisCaptureCheck.cpp. The code analysis approach wouldn't be perfect, but it might be good enough.

mattklein123 commented:
> One other design alternative: could we remove the need to lock a weak_ptr at every callback if, in startGlobalThreading, we go through all the callbacks and delete the ones that refer to a deleted slot? That would make a clear transition from "we're doing crazy stuff to handle startup-only problems" to "normal running state".

Yes this might be possible. Let me think about this more.

lambdai commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -17,93 +17,121 @@ thread_local InstanceImpl::ThreadLocalData InstanceImpl::thread_local_data_;
+}
+
+InstanceImpl::SlotImpl::SlotImpl(InstanceImpl& parent, uint64_t index)
+    : parent_(parent), index_(index), still_alive_guard_(std::make_shared<bool>()),
+      active_callback_guard_(
+          new bool,
```
1. nit: `std::make_shared<bool>(false);`  though I am not sure if `false` is needed. 
2. `new bool()` or `new bool(false)`. Otherwise the value is not initialized.

lambdai commented:
@mattklein123 IIRC http route config is published to worker by TLS. Route is probably carrying  amount update, this may be acceptable or may not.

If EDS cluster update  is going through TLS then this approach would impact a lot.

lambdai commented:
> @mattklein123 IIRC http route config is published to worker by TLS. Route is probably carrying amount update, this may be acceptable or may not.
> 
> If EDS cluster update is going through TLS then this approach would impact a lot.

I could be wrong. `publish by update` is not impacted since the slot is not destroyed within

stevenzzzz commented:
> @mattklein123 IIRC http route config is published to worker by TLS. Route is probably carrying amount update, this may be acceptable or may not.
> 
It's likely a problem in the case of SRDS relatively frequently updates its scopes, with ~100workers, and say deleting 10 scopes(routeConfigProviders), this could mean main thread block for up to (100 events in workers) x 10 times.

> If EDS cluster update is going through TLS then this approach

mattklein123 commented:
> It's likely a problem in the case of SRDS relatively frequently updates its scopes, with ~100workers, and say deleting 10 scopes(routeConfigProviders), this could mean main thread block for up to (100 events in workers) x 10 times.

Yes the blocking is the biggest risk with this approach. It's hard to say whether this is a real issue in practice or not. I'm skeptical it is but hard to say.

mattklein123 commented:
@ahedberg @lambdai @lizan @jmarantz @stevenzzzz @ggreenway I want to summarize some of the open lines of thought here so we can have a discussion on how to move forward. I don't want to spend more time until we decide how we want to handle this:

- This change will prevent many classes of bugs we have had around TLS slot lifetime, but has the large downside of forcing blocking on the main thread while TLS slot updates flush.
- It's possible we can get pretty far with a clang-tidy plugin that looks like this one: https://github.com/mesos/clang-tools-extra/blob/mesos_90/clang-tidy/mesos/ThisCaptureCheck.cpp. The main thing that concerns me about the plugin approach is that it's only a partial fix. a) depending on what is captured it might still be broken depending on transitive lifetimes. b) there are potential race conditions (e.g. let's say we capture by weak_ptr and lock(). Depending on the structure of transitive lifetimes, we may lock something but there is still a race with destruction on the main thread.
- The bookkeeper was already implemented by @stevenzzzz to fix similar problems (see https://docs.google.com/document/d/1_axmNmbkDO3aQE3jS5-WgSLvXyVxKwapOnrkBZn-DSk/edit#heading=h.ui18fonf1w1) but it doesn't fix all such problems and adds complexity. (basically the bookkeeper code keeps the slot from getting recycled until all callbacks are flushed, but still relies on the caller to not capture anything they are not supposed to be capturing).

I will admit I'm not thrilled w/ the blocking proposed in this PR, but it doesn't feel completely terrible to me, especially as it removes the bookkeeper logic and prevents more potential issues.

If we don't want the blocking and want to try for a more systemic fix beyond a clang-tidy plugin, one option that I have been noodling on is to figure out how to defer delete the entire listener, cluster, etc. when all nested TLS callbacks have flushed. This would be effectively extending the bookkeeper method on a larger scale, to prevent main thread objects from dying until all nested callbacks have flushed. I think with the way we pass around TLS slot factories, this may be less horrible to accomplish than it seems.

So to summarize:
1) Do nothing, fix one off bugs that show up.
2) Try for a clang-tidy plugin, fix one off bugs that show up.
3) Do this PR (or a variant of it).
4) Figure out how to extend bookkeeper to a larger level.

Thoughts?

stevenzzzz commented:
> @ahedberg @lambdai @lizan @jmarantz @stevenzzzz @ggreenway I want to summarize some of the open lines of thought here so we can have a discussion on how to move forward. I don't want to spend more time until we decide how we want to handle this:
> 
> * This change will prevent many classes of bugs we have had around TLS slot lifetime, but has the large downside of forcing blocking on the main thread while TLS slot updates flush.
> * It's possible we can get pretty far with a clang-tidy plugin that looks like this one: https://github.com/mesos/clang-tools-extra/blob/mesos_90/clang-tidy/mesos/ThisCaptureCheck.cpp. The main thing that concerns me about the plugin approach is that it's only a partial fix. a) depending on what is captured it might still be broken depending on transitive lifetimes. b) there are potential race conditions (e.g. let's say we capture by weak_ptr and lock(). Depending on the structure of transitive lifetimes, we may lock something but there is still a race with destruction on the main thread.
> * The bookkeeper was already implemented by @stevenzzzz to fix similar problems (see https://docs.google.com/document/d/1_axmNmbkDO3aQE3jS5-WgSLvXyVxKwapOnrkBZn-DSk/edit#heading=h.ui18fonf1w1) but it doesn't fix all such problems and adds complexity. (basically the bookkeeper code keeps the slot from getting recycled until all callbacks are flushed, but still relies on the caller to not capture anything they are not supposed to be capturing).
> 
> I will admit I'm not thrilled w/ the blocking proposed in this PR, but it doesn't feel completely terrible to me, especially as it removes the bookkeeper logic and prevents more potential issues.
> 
> If we don't want the blocking and want to try for a more systemic fix beyond a clang-tidy plugin, one option that I have been noodling on is to figure out how to defer delete the entire listener, cluster, etc. when all nested TLS callbacks have flushed. This would be effectively extending the bookkeeper method on a larger scale, to prevent main thread objects from dying until all nested callbacks have flushed. I think with the way we pass around TLS slot factories, this may be less horrible to accomplish than it seems.
> 
> So to summarize:
> 
> 1. Do nothing, fix one off bugs that show up.
> 2. Try for a clang-tidy plugin, fix one off bugs that show up.
> 3. Do this PR (or a variant of it).
> 4. Figure out how to extend bookkeeper to a larger level.
> 
> Thoughts?

Basically Bookkeeper stops a slotimpl from being tearing down when there are outgoing lambdas in workers(for https://github.com/envoyproxy/envoy/issues/7902). It also doesnt block when the destructor is called on main thread. 
This PR extend the life cycle coverage to part of captured objects by checking if the slotimpl destructor is called. 

I think we can extend the bookeeper with the update we have in this PR, that will give us a nonblocking, and safer slot impl.

On the other hand, Tls is used to do cross-thread communication, we should probably state clearly in the doc that only stateless data should be captured and passed around, the clang tool will help here as well.

mattklein123 commented:
> I think we can extend the bookeeper with the update we have in this PR, that will give us a nonblocking, and safer slot impl.

How? Can you expand? I don't see any way of making it safer on its own without without blocking.

ggreenway commented:
If we went the clang-tidy route, would that be sufficient to also remove the BookKeeper code?

mattklein123 commented:
> If we went the clang-tidy route, would that be sufficient to also remove the BookKeeper code?

No, I don't think so. The bookkeeper code is preventing the slot from getting recycled while callbacks are still in flight. Fundamentally it's the same root issue: we have callbacks in flight that refer to something on the main thread that is in the process of getting destroyed.

This is why I'm still personally voting for the blocking solution: it doesn't thrill me but my intuition is that the blocking won't cause any issues in practice, especially with a combination of timeouts on the wait and the deadlock detection code we already have.

stevenzzzz commented:
> How? Can you expand? I don't see any way of making it safer on its own without without blocking.

I am thinking let's add the "still_alive_guard_" to bookkeeper, and in the wrapped callback capture the weak_ptr, check still_alive_guard_ is still alive before calling the real callback. this is an i'd say huge improvement to the bookkeeper already.
for callbacks which has won the race on "destructing/checking still_alive_guard_", since the slotimpl is not deleted, the deferred deletion trick adds some safety to these winner callbacks. (same as this PR, it wont solve all the problems).

mattklein123 commented:
> I am thinking let's add the "still_alive_guard_" to bookkeeper, and in the wrapped callback capture the weak_ptr, check still_alive_guard_ is still alive before calling the real callback. this is an i'd say huge improvement to the bookkeeper already.
for callbacks which has won the race on "destructing/checking still_alive_guard_", since the slotimpl is not deleted, the deferred deletion trick adds some safety to these winner callbacks. (same as this PR, it wont solve all the problems).

Yes I agree this can't hurt, and I can add this, but it doesn't actually fix the underlying race condition, but I suppose it's better than nothing.

mattklein123 commented:
Status update: I'm going to work on cleaning up the existing thread_local_impl code (I think it can be simplified) and also adding the still alive guard per discussion with @stevenzzzz. This should fix another class of bugs (in particular LDS listener failure during initial startup). After that I will take a look at a clang-tidy plugin. This seems like the least contentious path forward right now.

stevenzzzz commented:
Thanks!

mattklein123 commented:
@ahedberg @lambdai @lizan @jmarantz @stevenzzzz @ggreenway I finally got back to this. See the updated PR title and description for what I did. Let me know what you think of this direction. I think this will be non-controversial. If so, I will add some more tests. Thank you!

jmarantz commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
-    auto wrapper = std::make_unique<Bookkeeper>(*this, std::move(slot));
-    slots_.push_back(wrapper->slot_.get());
-    return wrapper;
+    SlotPtr slot(new SlotImpl(*this, slots_.size()));
```
nit: `SlotPtr slot = std::make_unique<SlotImpl>(*this, slots_.size()));`

jmarantz commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
   free_slot_indexes_.pop_front();
   ASSERT(idx < slots_.size());
-  SlotImplPtr slot(new SlotImpl(*this, idx));
+  SlotPtr slot(new SlotImpl(*this, idx));
```
ditto

jmarantz commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
+ThreadLocalObjectSharedPtr InstanceImpl::SlotImpl::get() { return getWorker(index_); }
+
+void InstanceImpl::SlotImpl::runOnAllThreads(const UpdateCb& cb, Event::PostCb complete_cb) {
+  runOnAllThreads([cb, index = index_]() { setThreadLocal(index, cb(getWorker(index))); },
```
OK it took me a while to understand why you are capturing `index` specifically rather than `this`, aided by the comments on `still_alive_guard_` below. I think this might be a good place for some more comments about lifetime expectations and race protection, or maybe. a pointer to a canonical place with those comments.

jmarantz commented on source/common/thread_local/thread_local_impl.h:
```diff
@@ -69,35 +50,30 @@ class InstanceImpl : Logger::Loggable<Logger::Id::main>, public NonCopyable, pub
+    // The following is used to safely verify via weak_ptr that this slot is still alive. This
+    // does not prevent all races if a callback does not capture appropriately, but it does fix
+    // the common case of a slot destroyed immediately before anything is posted to a worker.
+    // TODO(mattklein123): Add clang-tidy analysis rule to check that "this" is not captured by
```
Ah this might relate to my comment above.

jmarantz reviewed (commented):
I didn't look in detail at the prior state but I think I understand your use of weak_ptr now to avoid a race.

Are you going to use the thread synchronizer to make a test?

stevenzzzz commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
+Event::PostCb InstanceImpl::SlotImpl::wrapCallback(Event::PostCb cb) {
+  // See the header file comments for still_alive_guard_ for the purpose of this capture.
+  return [still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
+    if (!still_alive_guard.expired()) {
```
is my understanding correct now that we dont defer deletion of slot in main thread, we lost the safe guard on cb() accessing the soon deleted slot after passing the still_alive_guard?

mattklein123 replied:
> is my understanding correct now that we dont defer deletion of slot in main thread, we lost the safe guard on cb() accessing the soon deleted slot after passing the still_alive_guard?

Sorry I can't full parse this. Rephrase? No deferred deletion is required anymore because we don't capture anything that is owned by the slot. Thus, the still alive guard will be immediately freed and any future checks on workers will fail. This doesn't prevent bad captures (see comments) and races, but this is no different from before. (It's better than before because callbacks will not run in man cases that have previously caused crashes.)

stevenzzzz replied:
sorry, I am thinking of this scenario: 
1. if (!still_alive_guard.expired()) passes.
2. cb() run in workers
3. slot destructs in main thread, (now still_alive_guard.expired). and cb access slot(), this leads to undefined behavior?

mattklein123 replied:
Why would the callback access the slot? I don't think this makes sense.

stevenzzzz replied:
yeah, we probably need some refactor to get them out of the cb. there are a few of them, for example: https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/cluster_manager_impl.cc#L654

mattklein123 replied:
OK thanks I will fix.

mattklein123 replied:
In order to fix this, I'm just removing the PostCb APIs, and switching the the UpdateCb style APIs which pass the current value, and it's actually forcing fixing a bunch of broken captures, so great suggestion!

/wait

stevenzzzz replied:
Thanks, I think the cleanup makes lots of sense, and much cleaner now!

mattklein123 commented:
> Are you going to use the thread synchronizer to make a test?

Assuming there are no objections to this approach I'm going to add a substantial number of new tests to cover this behavior.

ggreenway commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
+Event::PostCb InstanceImpl::SlotImpl::wrapCallback(Event::PostCb cb) {
+  // See the header file comments for still_alive_guard_ for the purpose of this capture.
+  return [still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
+    if (!still_alive_guard.expired()) {
```
If I understand correctly, the logic here is that if it is !expired() at the start of this operation, we don't care if it gets destructed concurrently with cb() running, because the destruction will get post()'d to this thread and thus won't actually destruct the thread-local state on this thread until cb() is done. Is that correct? Can you add more comments here?

mattklein123 replied:
Yes, correct, will add more comments.

ggreenway commented on source/common/thread_local/thread_local_impl.cc:
```diff
@@ -26,79 +26,73 @@ SlotPtr InstanceImpl::allocateSlot() {
-  parent_.runOnAllThreads([this, cb]() { setThreadLocal(index_, cb(get())); });
+Event::PostCb InstanceImpl::SlotImpl::wrapCallback(Event::PostCb cb) {
+  // See the header file comments for still_alive_guard_ for the purpose of this capture.
+  return [still_alive_guard = std::weak_ptr<bool>(still_alive_guard_), cb] {
```
can `cb` be `std::move()`d here for slight efficiency gain?

mattklein123 replied:
My assumption is the compiler will take care of this but let me see what I can do to make it more explicit. Adding more tests now.

mattklein123 commented:
@ggreenway @jmarantz @stevenzzzz updated per comments.

mattklein123 commented:
@ggreenway @jmarantz @stevenzzzz updated again. I fixed a bunch of broken captures and simplified the API. I will take another pass on broken captures when I work on the clang-tidy plugin.

ggreenway reviewed (approved):
I think this looks good. The code ended up simpler, no super-complicated relationship between threads/objects.

stevenzzzz reviewed (approved):
great cleanup!

jmarantz commented on source/common/stats/thread_local_store.cc:
```diff
@@ -206,11 +206,13 @@ void ThreadLocalStoreImpl::mergeHistograms(PostMergeCb merge_complete_cb) {
     tls_->runOnAllThreads(
-        [this]() -> void {
-          for (const auto& id_hist : tls_->getTyped<TlsCache>().tls_histogram_cache_) {
+        [](ThreadLocal::ThreadLocalObjectSharedPtr object)
```
A thought for a future revision, I think. For what I think is maybe nicer compile-time type safety, as well as less cumbersome usage, can we consider 2 changes:

1. Templatize the usage of slotes, so I put the fact that this is slotting a `TlsCache` into my declaration: `ThreadLocal::SlotPtr<TlsCache>` and then, for me, asType() is not needed.
2. Pass in object here as a non-const reference, so I can update if I like, but don't have to return it or declare the return-value.

Then this call-site becomes:

```
tls_->runOnAllThreads([](TlsCache& tls_cache) {
  for (const auto& id_hist : tls_cache.tls_histogram_cache_) {
    const TlsHistogramSharedPtr& tls_hist = id_hist.second;
    tls_hist->beginMerge();
  }
});
```

I think the Slot mechanism itself doesn't to change at all to accommodate this templating layer. It can do what it does now, and the templating API can be a thin header-only layer above.

mattklein123 replied:
Let me think about this more. I thought about the type of templating you suggest, but I think I decided it would get complicated with the APIs/callbacks. I can try to prototype it. I agree with the idea around passing a non-cost ref as a way of simplifying it. I'm going to open a follow up issue about adding clang-tidy and will add this also.

mattklein123 replied:
https://github.com/envoyproxy/envoy/issues/13313

jmarantz replied:
awesome, thanks! I'll follow along in that bug.

jmarantz reviewed (commented):
This is great, and a shout-out to well to @stevenzzzz for finding this race, as well as to Matt for fixing it in a very elegant way. One small suggestion. Could be a follow-up.
